src/Operation/GitSyncOperation.php

```diff
@@ -0,0 +1,181 @@
+
+        $this->gitHelper->remoteUpdate($this->sourceRemote);
+        $this->gitHelper->remoteUpdate($this->destinationRemote);
+
+        $this->gitHelper->guardBranchExist($this->localBranch);
+        $this->gitHelper->guardRemoteBranchExists($this->sourceRemote, $this->sourceBranch);
+
+        $count = $this->processHelper->runCommand(
+            [
+                'git',
+                'rev-list',
+                $this->localBranch.'...'.$this->sourceRemote.'/'.$this->sourceBranch,
+                '--count',
+            ]
+        );
+
+        // All up-to-date, nothing to do.
+        if (0 === $count) {
+            return;
+        }
+
+        $this->gitHelper->checkout($this->localBranch);
+
+        if (self::SYNC_FORCE === $type) {
+            $this->gitHelper->reset($this->sourceRemote.'/'.$this->sourceBranch, 'hard');
+
+            return;
+        }
+
+        if (self::SYNC_SMART !== $type && self::SYNC_SMART_MERGE !== $type) {
+            throw new \InvalidArgumentException('Unsupported sync option provided: '.$type);
+        }
+
+        switch ($this->gitHelper->getRemoteDiffStatus()) {
+            case 'up-to-date':
+                return;
+
+            case GitHelper::STATUS_NEED_PULL:
+                if (self::SYNC_SMART === $type) {
+                    $this->gitHelper->createRebaseOperation(
+                        $this->sourceRemote.'/'.$this->sourceBranch
+                    )->performRebase();
+                } else {
+                    $this->gitHelper->mergeBranch(
+                        $this->sourceRemote.'/'.$this->sourceBranch,
+                        $this->localBranch,
+                        '',
+                        true
+                    );
+                }
+                break;
+
+            case 'need-push':
+                $this->pushToRemote();
+                break;
+
+            case 'need-rebase-push':
+                if (!$allowForcePush) {
+                    throw new \RuntimeException(
+                        sprintf('Push with force is required but not enabled for the operation.')
+                    );
+                }
+
+                $baseRef = $this->processHelper->runCommand(
+                    ['git', 'rev-parse', $this->sourceRemote, $this->sourceBranch]
+                );
+
+                // Rebase onto the the current remote base to ensure the commits don't conflict.
+                $this->gitHelper->createRebaseOperation($baseRef)->onto(
+                    $this->sourceRemote.'/'.$this->sourceBranch,
+                    $baseRef,
+                    $this->localBranch
+                )->performRebase();
+
+                $this->pushToRemote(true);
+                break;
+        }
+    }
+
+    private function pushToRemote($force = false)
+    {
+        $target = trim($this->localBranch).':'.$this->destinationBranch;
+
+        // Safety guard to prevent deleting a remote base branch!!
+        if (':' === $target[0]) {
```

Comment:
I'm thinking about moving this safe-guard to pushToRemote() and then adding a new removeRemoteBranch() method for deleting remote branches :+1: Will do this in a separate PR with some other improvements.